lista3/src/Functions.scala

```diff
@@ -0,0 +1,170 @@
+      }
+      else splitIter(list.tail, negative, negativeAndOdd)
+    }
+    splitIter(list, Nil, Nil)
+  }
+
+  //zadanie 2
+  def length[A](list: List[A]): Int = {
+    @tailrec
+    def lengthIter(list: List[A], length: Int): Int = {
+      if (list == Nil) length
+      else lengthIter(list.tail, length + 1)
+    }
+    lengthIter(list, 0)
+  }
+
+  //zadanie 3
+  //zlozonosc obliczeniowa: O(2*(firstList.length + secondList.length)); zlozonosc pamieciowa: stala
+  def merge[A](firstList: List[A], secondList: List[A]): List[A] = {
+    @tailrec
+    def mergeIter(firstList: List[A], secondList: List[A], returnList: List[A]): List[A] = {
+      (firstList, secondList) match {
+        case (Nil, Nil) => reverse(returnList)
+        case (head :: tail, Nil) => mergeIter(tail, secondList, head :: returnList)
+        case (Nil, head :: tail) => mergeIter(firstList, tail, head :: returnList)
+        case (head :: tail, _) => mergeIter(secondList, tail, head :: returnList)
+      }
+    }
+    mergeIter(firstList, secondList, Nil)
+  }
+
+  //zadanie 4
+  @tailrec
+  private def matchStringToElementHelper(argument: String, element: String): Boolean = {
+    (argument, element) match {
+      case ("", "") => return true
+      case ("", _) => return false
+      case (_, "") => return true
+      case (_, _) =>
+        if (argument.head == element.head && stringLength(element) == 1) return true
+        else if (argument.head == element.head && stringLength(argument) > 1 && stringLength(element) > 1 && argument.tail.head == element.tail.head) matchStringToElementHelper(argument.tail.tail, element.tail.tail)
+        else matchStringToElementHelper(argument.tail, element)
+    }
+  }
+
+  private def matchStringToElement(argument: String, element: String): Boolean = {
+    if (argument.length < element.length) return false;
+    else matchStringToElementHelper(argument, element)
+  }
+
+  def find_Tail(list: List[String], element: String): List[String] = {
+    @tailrec
+    def find_TailIter(list: List[String], element: String, returnList: List[String]): List[String] = {
+      if (list == Nil) reverse(returnList)
+      else if (matchStringToElement(list.head, element)) find_TailIter(list.tail, element, list.head :: returnList)
+      else find_TailIter(list.tail, element, returnList)
+    }
+    find_TailIter(list, element, Nil)
+  }
+
+  def findEach_Tail(list: List[String], elements: List[String]): List[String] = {
+    @tailrec
+    def findEach_Tail_Iter(list: List[String], elements: List[String], returnList: List[String]): List[String] = {
+      if (elements == Nil) returnList
+      else {
+        @tailrec
+        def addMatchesToReturnList(matches: List[String], returnList: List[String]): List[String] = {
+          if (matches == Nil) returnList
+          else if (!contains(returnList, matches.head)) addMatchesToReturnList(matches.tail, matches.head :: returnList)
```

Comment:
zamiast sprawdzać czy już dodaliśmy można by użyć struktury danych, która nie pozwala na dodanie duplikatu. Poza tym co jesli bedziemy wyszukiwać trzy takie same slowa przyklad "bella ciao ciao ciao" chcemy dostać zwrotke taka jak szukalismy czy obciete do "bella ciao"? Obstawiam to pierwsze.